packages/datadog-shimmer/src/shimmer.js

```diff
@@ -142,14 +123,24 @@ function safeWrapper (original, wrapper) {
   // managing the callState, but then we'd be calling `wrapper` on each invocation, so
   // instead we do it here, once.
   const innerWrapped = wrapper(function (...args) {
+    // This is only used in safe mode. It's a simple state machine to track if the
+    // original method was called and if it returned. We need this to determine if
+    // an error was thrown by the original method, or by us. We'll use one of these
+    // per call to a wrapped method.
     // We need to stash the callState here because of recursion.
     const callState = currentCallState
-    callState.startCall()
+    callState[0] = 1
     const retVal = original.apply(this, args)
     if (isPromise(retVal)) {
-      retVal.then(callState.endCall.bind(callState))
+      // TODO: should this not be consistently be either a promise or not? I
+      // would expect we only have to check once and can safe that state.
+      retVal.then((value) => {
+        callState[0] = 2
+        callState[1] = value
```

Comment:
so from what I understand:
`callState[0] = 0` -> not called
`callState[0] = 1` -> called
`callState[0] = 2` -> completed
`callState[1]` -> return value
is that correct ?
It doesn't look like this is explained anywhere? I had to squint my eyes and use my brain to reverse engineer it.

Two suggestions:
1. something like: `callState[STATE] = COMPLETED`
2. a good comment block to explain all that

that shouldn't decrease the optimization right ?